File: 443.string-compression.py
#
# @lc app=leetcode id=443 lang=python3
#
# [443] String Compression
#
from typing import List

# @lc code=start
class Solution:
    # T: O(N)
    # S: O(1)
    def compress(self, chars: List[str]) -> int:
        N = len(chars)
        i = j = 0

        while i < N:
            cur = chars[i]
            cnt = 1
            i += 1
            while i < N and chars[i] == cur:
                i += 1
                cnt += 1
            
            chars[j] = cur
            j += 1

            if cnt > 1:

                # If cnt > 10, need to reverse the digits
                k1 = j
                while cnt != 0:
                    chars[j] = str(cnt % 10)
                    cnt //= 10
                    j += 1
                k2 = j - 1
                
                # Below is O(1) space and faster than chars[k1:j] = reversed(chars[k1:j])
                while k1 < k2:
                    chars[k1], chars[k2] = chars[k2], chars[k1]
                    k1 += 1
                    k2 -= 1

        return j
        
# @lc code=end

File: 443.string-compression.cpp
/*
 * @lc app=leetcode id=443 lang=cpp
 *
 * [443] String Compression
 */
#include <string>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
public:
    // T: O(n)
    // S: O(1)
    int compress(vector<char>& chars) {
        const int n = chars.size();
        int i = 0, j = 0;
        while (i < n) {
            char cur = chars[i];
            int cnt = 0;
            while (i < n && chars[i] == cur) {
                ++i; 
                ++cnt;
            }
            chars[j++] = cur;
            if (cnt != 1) {
                // len(str(cnt)) is always no more than the number of repeated characters
                /*
                // Below takes O(log10(n)) space
                for (const auto &c : to_string(cnt)) {
                    chars[j++] = c;
                }
                */

                // Below keeps space complexity O(1) and seems to be faster,
                // even though needs to reverse the digits
                int k = j;
                while (cnt) {
                    chars[j++] = (char) ((cnt % 10) + '0'); 
                    cnt /= 10;
                }
                reverse(chars.begin() + k, chars.begin() + j);
            }
        }
        return j;
    }
};
// @lc code=end

